File: Php/Infrastructure/Data/MySql/MySqlSharingStorage.php
namespace Infrastructure\Data\MySql {

	use Infrastructure\Data\Interfaces\SharingStorage;
	use Infrastructure\Data\MySql\StorageProvider;
	use Domain\Classes\User;

	class MySqlSharingStorage implements SharingStorage{

		private $db;

		public function __construct(StorageProvider $db){
			db = $db;
		}

		public function getServices(User $user){
			$sql =	"SELECT us.ServiceName " .
					"FROM usersharing us " .
					"WHERE us.UserID = {$user.getId()}";
			return getServicesFromSql($sql);
		}

		public function isSharing(User $user, $sharingProvider){
			$sql =	"SELECT us.UserID, us.ServiceName " .
					"FROM usersharing us " .
					"WHERE us.UserID = {$user.getId()} " .
					"AND us.ServiceName = '{$sharingProvider}'";
			return getSharingStatusFromSql($sql);
		}

		public function addSharing(User $user, $sharingProvider){
			$sql =	"INSERT INTO usersharing " .
					"(UserID, ServiceName) " .
					"VALUES " .
					"({$user.getId()}, '{$sharingProvider}')";
			db.execute($sql);
		}

		public function removeSharing(User $user, $sharingProvider){
			$sql =	"DELETE FROM usersharing " .
					"WHERE UserID = {$user.getId()} " .
					"AND ServiceName = '{$sharingProvider}'";
			$rowCount = db.execute($sql);
			return $rowCount > 0;
		}

		private function getServicesFromSql($sql){
			$res = db.query($sql);
			$services = array();
			foreach($res.fetchAll() as $row){
				$services[] = serviceFromDbRow($row);
			}
			return $services;
		}

		private function getSharingStatusFromSql($sql){
			$res = db.query($sql);
			foreach($res.fetchAll() as $row){
				return true;
				break;
			}
			return false;
		}

		private function serviceFromDbRow($row){
			return $row["ServiceName"];
		}

	}

}

File: Infrastructure/Data/Storage/MySqlSharingStorage.cs
using System.Collections.Generic;
using Infrastructure.Data.Storage.Interfaces;

namespace Infrastructure.Data.Storage {

	public class MySqlSharingStorage : ISharingStorage
    {
	    private readonly IStorageProvider _storageProvider;

	    public MySqlSharingStorage(IStorageProvider storageProvider)
	    {
	        _storageProvider = storageProvider;
	    }

	    public IList<string> GetServices(int userId){
			var sql = "SELECT us.ServiceName FROM usersharing us WHERE us.UserID = {0}";
	        sql = string.Format(sql, userId);
			return GetServicesFromSql(sql);
		}

		public bool IsSharing(int userId, string sharingProvider){
			var sql = "SELECT us.UserID, us.ServiceName FROM usersharing us WHERE us.UserID = {0} AND us.ServiceName = '{1}'";
		    sql = string.Format(sql, userId, sharingProvider);
			return GetSharingStatusFromSql(sql);
		}

		public void AddSharing(int userId, string sharingProvider){
			var sql = "INSERT INTO usersharing (UserID, ServiceName) VALUES ({0}, '{1}')";
            sql = string.Format(sql, userId, sharingProvider);
			_storageProvider.ExecuteInsert(sql);
		}

		public void RemoveSharing(int userId, string sharingProvider){
			var sql = "DELETE FROM usersharing WHERE UserID = {0} AND ServiceName = '{1}'";
			sql = string.Format(sql, userId, sharingProvider);
            _storageProvider.Execute(sql);
		}

		private IList<string> GetServicesFromSql(string sql){
            var reader = _storageProvider.Query(sql);
		    var services = new List<string>();
			while(reader.Read()){
				services.Add(reader.GetString("ServiceName"));
			}
			return services;
		}

		private bool GetSharingStatusFromSql(string sql){
			var reader = _storageProvider.Query(sql);
            return reader.Read();
		}

    }

}